fl4health/model_bases/feature_extractor_buffer.py

```diff
@@ -0,0 +1,187 @@
+
+    def find_last_common_prefix(self, prefix: str, layers_name: List[str]) -> str:
+        """
+        Check the model's list of named modules to filter any layer that starts with the given prefix and
+        return the last one.
+
+        Args:
+            prefix (str): The prefix of the layer name for registering the hook.
+            layers_name (List[str]): The list of named modules of the model. The assumption is that list of
+            named modules is sorted in the order of the model's forward pass with depth-first traversal. This
+            will allow the user to specify the generic name of the layer instead of the full hierarchical name.
+
+        Returns:
+            str: The complete name of last named layer that matches the prefix.
+        """
+        filtered_layers = [layer for layer in layers_name if layer.startswith(prefix)]
+
+        # Return the last element that matches the criteria
+        return filtered_layers[-1]
+
+    def _maybe_register_hooks(self) -> None:
+        """
+        Checks if hooks are already registered and registers them if not.
+        Hooks extract the intermediate feature as output of the selected layers in the model.
+        """
+        if len(self.fhooks) == 0:
+            log(INFO, "Starting to register hooks:")
+            named_layers = list(dict(self.model.named_modules()).keys())
+            for layer in self.flatten_feature_extraction_layers.keys():
+                log(INFO, f"Registering hook for layer: {layer}")
+                # Find the last specific layer under a given generic name
+                specific_layer = self.find_last_common_prefix(layer, named_layers)
+                # Split the specific layer name by '.' to get the hierarchical attribute
+                layer_hierarchicy_list = specific_layer.split(".")
+                self.fhooks.append(
+                    self.get_hierarchical_attr(self.model, layer_hierarchicy_list).register_forward_hook(
+                        self.forward_hook(layer)
+                    )
+                )
+        else:
+            log(INFO, "Hooks already registered.")
+
+    def remove_hooks(self) -> None:
+        """
+        Removes the hooks from the model for checkpointing and clears the hook list. This method is used to remove
+        any hooks that have been added to the feature extractor buffer. It is typically called prior to checkpointing
+        the model.
+        """
+
+        log(INFO, "Removing hooks.")
+        for hook in self.fhooks:
+            hook.remove()
+        self.fhooks.clear()
+
+    def forward_hook(self, layer_name: str) -> Callable:
+        """
+        Returns a hook function that is called during the forward pass of a module.
+
+        Args:
+            layer_name (str): The name of the layer.
+
+        Returns:
+            Callable: The hook function that takes in a module, input, and output tensors.
+
```

Comment:
Very minor and maybe just a github rendering thing, but is there an extra endline here?